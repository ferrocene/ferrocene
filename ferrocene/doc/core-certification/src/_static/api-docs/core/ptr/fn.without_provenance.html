<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Creates a pointer with the given address and no provenance."><title>without_provenance in core::ptr - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="core" data-themes="" data-resource-suffix="1.88.0" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../../static.files/storage-82c7156e.js"></script><script defer src="sidebar-items1.88.0.js"></script><script defer src="../../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc fn"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../core/index.html"><img class="rust-logo" src="../../static.files/rust-logo-9a9549ea.svg" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../core/index.html"><img class="rust-logo" src="../../static.files/rust-logo-9a9549ea.svg" alt="logo"></a><h2><a href="../../core/index.html">core</a><span class="version">1.88.0-dev</span></h2></div><div class="sidebar-elems"><div id="rustdoc-modnav"><h2><a href="index.html">In core::<wbr>ptr</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">core</a>::<wbr><a href="index.html">ptr</a></div><h1>Function <span class="fn">without_provenance</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><span class="since" title="Stable since Rust version 1.84.0, const since 1.84.0">1.84.0 (const: 1.84.0)</span> Â· <a class="src" href="../../src/core/ptr/mod.rs.html#621-623">Source</a> </span></div><pre class="rust item-decl"><code>pub const fn without_provenance&lt;T&gt;(addr: <a class="primitive" href="../primitive.usize.html">usize</a>) -&gt; *const T</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Creates a pointer with the given address and no <a href="index.html#provenance" title="mod core::ptr">provenance</a>.</p>
<p>This is equivalent to <code>ptr::null().with_addr(addr)</code>.</p>
<p>Without provenance, this pointer is not associated with any actual allocation. Such a
no-provenance pointer may be used for zero-sized memory accesses (if suitably aligned), but
non-zero-sized memory accesses with a no-provenance pointer are UB. No-provenance pointers are
little more than a <code>usize</code> address in disguise.</p>
<p>This is different from <code>addr as *const T</code>, which creates a pointer that picks up a previously
exposed provenance. See [<code>with_exposed_provenance</code>] for more details on that operation.</p>
<p>This is a <a href="index.html#strict-provenance" title="mod core::ptr">Strict Provenance</a> API.</p>
</div></details></section></div></main></body></html>