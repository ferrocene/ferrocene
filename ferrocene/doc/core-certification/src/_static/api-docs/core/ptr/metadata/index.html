<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `metadata` mod in crate `core`."><title>core::ptr::metadata - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="core" data-themes="" data-resource-suffix="1.88.0" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../../../static.files/storage-82c7156e.js"></script><script defer src="../sidebar-items1.88.0.js"></script><script defer src="../../../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../../core/index.html"><img class="rust-logo" src="../../../static.files/rust-logo-9a9549ea.svg" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../../core/index.html"><img class="rust-logo" src="../../../static.files/rust-logo-9a9549ea.svg" alt="logo"></a><h2><a href="../../../core/index.html">core</a><span class="version">1.88.0</span></h2></div><div class="version">(e925a16bd	2025-08-18)</div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module metadata</a></h2><h3><a href="#traits">Module Items</a></h3><ul class="block"><li><a href="#traits" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#trait-aliases" title="Trait Aliases">Trait Aliases</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In core::<wbr>ptr</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">core</a>::<wbr><a href="../index.html">ptr</a></div><h1>Module <span>metadata</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/core/ptr/metadata.rs.html#1-287">Source</a> </span></div><span class="item-info"><div class="stab unstable"><span class="emoji">🔬</span><span>This is a nightly-only experimental API. (<code>ptr_metadata</code>&nbsp;<a href="https://github.com/rust-lang/rust/issues/81513">#81513</a>)</span></div></span><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.Pointee.html" title="trait core::ptr::metadata::Pointee">Pointee</a><wbr><span class="stab unstable" title="">Experimental</span></dt><dd>Provides the pointer metadata type of any pointed-to type.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.from_raw_parts.html" title="fn core::ptr::metadata::from_raw_parts">from_<wbr>raw_<wbr>parts</a><wbr><span class="stab unstable" title="">Experimental</span></dt><dd>Forms a (possibly-wide) raw pointer from a data pointer and metadata.</dd><dt><a class="fn" href="fn.from_raw_parts_mut.html" title="fn core::ptr::metadata::from_raw_parts_mut">from_<wbr>raw_<wbr>parts_<wbr>mut</a><wbr><span class="stab unstable" title="">Experimental</span></dt><dd>Performs the same functionality as <a href="fn.from_raw_parts.html" title="fn core::ptr::metadata::from_raw_parts"><code>from_raw_parts</code></a>, except that a
raw <code>*mut</code> pointer is returned, as opposed to a raw <code>*const</code> pointer.</dd></dl><h2 id="trait-aliases" class="section-header">Trait Aliases<a href="#trait-aliases" class="anchor">§</a></h2><dl class="item-table"><dt><a class="traitalias" href="traitalias.Thin.html" title="traitalias core::ptr::metadata::Thin">Thin</a><wbr><span class="stab unstable" title="">Experimental</span></dt><dd>Pointers to types implementing this trait alias are “thin”.</dd></dl></section></div></main></body></html>