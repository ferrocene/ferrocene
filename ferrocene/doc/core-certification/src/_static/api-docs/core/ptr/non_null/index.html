<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `non_null` mod in crate `core`."><title>core::ptr::non_null - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../../static.files/rustdoc-916cea96.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="core" data-themes="" data-resource-suffix="1.88.0" data-rustdoc-version="1.87.0 (17067e9ac 2025-05-09)" data-channel="1.87.0" data-search-js="search-e7298875.js" data-settings-js="settings-d72f25bb.js" ><script src="../../../static.files/storage-82c7156e.js"></script><script defer src="../sidebar-items1.88.0.js"></script><script defer src="../../../static.files/main-fb8c74a8.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../../core/index.html"><img class="rust-logo" src="../../../static.files/rust-logo-9a9549ea.svg" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../../core/index.html"><img class="rust-logo" src="../../../static.files/rust-logo-9a9549ea.svg" alt="logo"></a><h2><a href="../../../core/index.html">core</a><span class="version">1.88.0-dev</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module non_<wbr>null</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="../index.html">In core::<wbr>ptr</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../../index.html">core</a>::<wbr><a href="../index.html">ptr</a></div><h1>Module <span>non_null</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../../src/core/ptr/non_null.rs.html#1-1707">Source</a> </span></div><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.NonNull.html" title="struct core::ptr::non_null::NonNull">NonNull</a></dt><dd><code>*mut T</code> but non-zero and <a href="https://doc.rust-lang.org/reference/subtyping.html">covariant</a>.</dd></dl></section></div></main></body></html>