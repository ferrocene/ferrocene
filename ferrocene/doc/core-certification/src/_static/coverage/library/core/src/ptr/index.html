<!DOCTYPE html>
<html lang="en-us">
    <head><meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <title>Grcov report - library&#x2F;core&#x2F;src&#x2F;ptr </title>
        <link rel="stylesheet" href="../../../../bulma.min.css"></head>
    <body>
        <div class="container">
    <nav class="breadcrumb is-right" aria-label="breadcrumbs">
        <ul><li><a href="../../../../index.html">top_level</a></li><li class="is-active"><a href="#">library&#x2F;core&#x2F;src&#x2F;ptr</a></li>
        </ul>
    </nav>
    <nav class="level">
        <div class="level-item has-text-centered">
        <div>
            <p class="heading">Lines</p>
            <p class="title has-text-success">
                <abbr title="68 / 68">100 %</abbr></p>
        </div>
    </div>

        <div class="level-item has-text-centered">
        <div>
            <p class="heading">Functions</p>
            <p class="title has-text-warning">
                <abbr title="555 / 719">77.19 %</abbr></p>
        </div>
    </div>

        
    </nav>

    <table class="table is-fullwidth">
        <thead>
            <tr>
                <th>File</th>
                <th class="has-text-centered" colspan="3">Line Coverage</th>
                <th class="has-text-centered" colspan="2">Functions</th>
                
            </tr>
        </thead>
        <tbody>
                    
    <tr>
        <th><a href="alignment.rs.html">alignment.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            13 / 13
        </td>
        <!-- -->
        <td class="has-text-centered has-background-success p-2">100%</td>
        <td class="has-text-centered has-background-success p-2">3 / 3 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="const_ptr.rs.html">const_ptr.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            16 / 16
        </td>
        <!-- -->
        <td class="has-text-centered has-background-warning p-2">79.39%</td>
        <td class="has-text-centered has-background-warning p-2">131 / 165 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="metadata.rs.html">metadata.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            12 / 12
        </td>
        <!-- -->
        <td class="has-text-centered has-background-warning p-2">78.4%</td>
        <td class="has-text-centered has-background-warning p-2">127 / 162 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="mod.rs.html">mod.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            24 / 24
        </td>
        <!-- -->
        <td class="has-text-centered has-background-danger p-2">74.49%</td>
        <td class="has-text-centered has-background-danger p-2">184 / 247 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="mut_ptr.rs.html">mut_ptr.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            3 / 3
        </td>
        <!-- -->
        <td class="has-text-centered has-background-warning p-2">77.46%</td>
        <td class="has-text-centered has-background-warning p-2">110 / 142 </td>
        <!-- -->
        
    </tr>

                  </tbody>
    </table></div>
        <footer class="footer">
            
        </footer>
    </body>
</html>
