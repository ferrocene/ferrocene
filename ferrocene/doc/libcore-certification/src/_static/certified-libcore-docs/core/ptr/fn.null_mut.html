<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Creates a null mutable raw pointer."><title>null_mut in core::ptr - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-84e720fa.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="core" data-themes="" data-resource-suffix="1.89.0" data-rustdoc-version="1.89.0-dev" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="sidebar-items1.89.0.js"></script><script defer src="../../static.files/main-fd3af306.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-32bb7600.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc fn"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../core/index.html"><img class="rust-logo" src="../../static.files/rust-logo-9a9549ea.svg" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../core/index.html"><img class="rust-logo" src="../../static.files/rust-logo-9a9549ea.svg" alt="logo"></a><h2><a href="../../core/index.html">core</a><span class="version">1.89.0-dev</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">null_<wbr>mut</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#examples" title="Examples">Examples</a></li></ul></section><div id="rustdoc-modnav"><h2><a href="index.html">In core::<wbr>ptr</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">core</a>::<wbr><a href="index.html">ptr</a></div><h1>Function <span class="fn">null_mut</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><span class="since" title="Stable since Rust version 1.6.0, const since 1.24.0">1.6.0 (const: 1.24.0)</span> · <a class="src" href="../../src/core/ptr/mod.rs.html#877-879">Source</a> </span></div><pre class="rust item-decl"><code>pub const fn null_mut&lt;T: ?<a class="trait" href="../marker/trait.Sized.html" title="trait core::marker::Sized">Sized</a> + <a class="traitalias" href="metadata/traitalias.Thin.html" title="traitalias core::ptr::metadata::Thin">Thin</a>&gt;() -&gt; *mut T</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Creates a null mutable raw pointer.</p>
<p>This function is equivalent to zero-initializing the pointer:
<code>MaybeUninit::&lt;*mut T&gt;::zeroed().assume_init()</code>.
The resulting pointer has the address 0.</p>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>std::ptr;

<span class="kw">let </span>p: <span class="kw-2">*mut </span>i32 = ptr::null_mut();
<span class="macro">assert!</span>(p.is_null());
<span class="macro">assert_eq!</span>(p <span class="kw">as </span>usize, <span class="number">0</span>); <span class="comment">// this pointer has the address 0</span></code></pre><a class="test-arrow" target="_blank" title="Run code" href="https://play.rust-lang.org/?code=%23!%5Ballow(unused)%5D%0Afn+main()+%7B%0A++++use+std::ptr;%0A++++%0A++++let+p:+*mut+i32+=+ptr::null_mut();%0A++++assert!(p.is_null());%0A++++assert_eq!(p+as+usize,+0);%0A%7D&amp;edition=2024"></a></div>
</div></details></section></div></main></body></html>