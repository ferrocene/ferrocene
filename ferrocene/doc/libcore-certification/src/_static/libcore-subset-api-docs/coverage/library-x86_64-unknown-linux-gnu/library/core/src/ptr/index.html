<!DOCTYPE html>
<html lang="en-us">
    <head><meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <title>Grcov report - library&#x2F;core&#x2F;src&#x2F;ptr </title>
        <link rel="stylesheet" href="../../../../bulma.min.css"></head>
    <body>
        <div class="container">
    <nav class="breadcrumb is-right" aria-label="breadcrumbs">
        <ul><li><a href="../../../../index.html">top_level</a></li><li class="is-active"><a href="#">library&#x2F;core&#x2F;src&#x2F;ptr</a></li>
        </ul>
    </nav>
    <nav class="level">
        <div class="level-item has-text-centered">
        <div>
            <p class="heading">Lines</p>
            <p class="title has-text-danger">
                <abbr title="54 / 75">72 %</abbr></p>
        </div>
    </div>

        <div class="level-item has-text-centered">
        <div>
            <p class="heading">Functions</p>
            <p class="title has-text-danger">
                <abbr title="144 / 469">30.7 %</abbr></p>
        </div>
    </div>

        
    </nav>

    <table class="table is-fullwidth">
        <thead>
            <tr>
                <th>File</th>
                <th class="has-text-centered" colspan="3">Line Coverage</th>
                <th class="has-text-centered" colspan="2">Functions</th>
                
            </tr>
        </thead>
        <tbody>
                    
    <tr>
        <th><a href="alignment.rs.html">alignment.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-warning is-large"
                value="84.61538461538461"
                max="100">
                84.62%
            </progress>
        </td>
        <td class="has-text-centered has-background-warning p-2">
            84.62%
        </td>
        <td class="has-text-centered has-background-warning p-2">
            11 / 13
        </td>
        <!-- -->
        <td class="has-text-centered has-background-success p-2">100%</td>
        <td class="has-text-centered has-background-success p-2">3 / 3 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="const_ptr.rs.html">const_ptr.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-danger is-large"
                value="34.48275862068966"
                max="100">
                34.48%
            </progress>
        </td>
        <td class="has-text-centered has-background-danger p-2">
            34.48%
        </td>
        <td class="has-text-centered has-background-danger p-2">
            10 / 29
        </td>
        <!-- -->
        <td class="has-text-centered has-background-danger p-2">32.53%</td>
        <td class="has-text-centered has-background-danger p-2">27 / 83 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="metadata.rs.html">metadata.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            12 / 12
        </td>
        <!-- -->
        <td class="has-text-centered has-background-danger p-2">29.77%</td>
        <td class="has-text-centered has-background-danger p-2">39 / 131 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="mod.rs.html">mod.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            18 / 18
        </td>
        <!-- -->
        <td class="has-text-centered has-background-danger p-2">26.37%</td>
        <td class="has-text-centered has-background-danger p-2">48 / 182 </td>
        <!-- -->
        
    </tr>

                  
                    
    <tr>
        <th><a href="mut_ptr.rs.html">mut_ptr.rs</a></th>
        <!-- -->
        <td class="p-2">
            <progress
                class="progress is-success is-large"
                value="100"
                max="100">
                100%
            </progress>
        </td>
        <td class="has-text-centered has-background-success p-2">
            100%
        </td>
        <td class="has-text-centered has-background-success p-2">
            3 / 3
        </td>
        <!-- -->
        <td class="has-text-centered has-background-danger p-2">38.57%</td>
        <td class="has-text-centered has-background-danger p-2">27 / 70 </td>
        <!-- -->
        
    </tr>

                  </tbody>
    </table></div>
        <footer class="footer">
            
        </footer>
    </body>
</html>
