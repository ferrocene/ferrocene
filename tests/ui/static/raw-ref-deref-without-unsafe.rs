use std::ptr;

// This code should remain unsafe because of the two unsafe operations here,
// even if in a hypothetical future we deem all &raw (const|mut) *ptr exprs safe.

static mut BYTE: u8 = 0;
static mut BYTE_PTR: *mut u8 = ptr::addr_of_mut!(BYTE);
// An unsafe static's ident is a place expression in its own right, so despite the above being safe
// (it's fine to create raw refs to places!) the following derefs the ptr before creating its ref!
static mut DEREF_BYTE_PTR: *mut u8 = ptr::addr_of_mut!(*BYTE_PTR);
//~^ ERROR: use of mutable static
//~| ERROR: dereference of raw pointer

fn main() {
    let _ = unsafe { DEREF_BYTE_PTR };
}
