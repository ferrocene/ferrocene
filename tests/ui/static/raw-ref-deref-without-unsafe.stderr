error[E0133]: dereference of raw pointer is unsafe and requires unsafe function or block
  --> $DIR/raw-ref-deref-without-unsafe.rs:10:56
   |
LL | static mut DEREF_BYTE_PTR: *mut u8 = ptr::addr_of_mut!(*BYTE_PTR);
   |                                                        ^^^^^^^^^ dereference of raw pointer
   |
   = note: raw pointers may be null, dangling or unaligned; they can violate aliasing rules and cause data races: all of these are undefined behavior

error[E0133]: use of mutable static is unsafe and requires unsafe function or block
  --> $DIR/raw-ref-deref-without-unsafe.rs:10:57
   |
LL | static mut DEREF_BYTE_PTR: *mut u8 = ptr::addr_of_mut!(*BYTE_PTR);
   |                                                         ^^^^^^^^ use of mutable static
   |
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0133`.
